File: concierge/utils/date_utils.py
"""
Date utility functions for consistent date handling across the application.

This module provides centralized date handling to ensure consistency between:
- iCal parsing and storage
- Firestore and DynamoDB storage
- UI display and JavaScript handling
- Conversation data mapping
"""

from datetime import datetime, date, timezone
from typing import Union, Optional
import re


def to_date_only(date_input: Union[datetime, date, str]) -> str:
    """
    Convert any date input to a date-only string format (YYYY-MM-DD).
    
    This function ensures all dates are stored and handled consistently
    as date-only strings, avoiding timezone-related issues.
    
    Args:
        date_input: datetime object, date object, or ISO string
        
    Returns:
        Date string in YYYY-MM-DD format
        
    Raises:
        ValueError: If the input cannot be parsed as a date
    """
    if date_input is None:
        raise ValueError("Date input cannot be None")
    
    if isinstance(date_input, str):
        # Handle ISO format strings
        if 'T' in date_input:
            # Parse datetime string and extract date part
            try:
                dt = datetime.fromisoformat(date_input.replace('Z', '+00:00'))
                return dt.date().isoformat()
            except ValueError:
                # Try parsing without timezone info
                dt = datetime.fromisoformat(date_input.split('T')[0])
                return dt.date().isoformat()
        else:
            # Already a date string, validate format
            try:
                parsed_date = datetime.strptime(date_input, '%Y-%m-%d').date()
                return parsed_date.isoformat()
            except ValueError:
                raise ValueError(f"Invalid date string format: {date_input}")
    
    elif isinstance(date_input, datetime):
        return date_input.date().isoformat()
    
    elif isinstance(date_input, date):
        return date_input.isoformat()
    
    else:
        raise ValueError(f"Unsupported date input type: {type(date_input)}")


def parse_ical_date(ical_date: Union[datetime, date]) -> str:
    """
    Parse iCal date/datetime and return as date-only string.
    
    Args:
        ical_date: Date or datetime from iCal parsing
        
    Returns:
        Date string in YYYY-MM-DD format
    """
    if isinstance(ical_date, date) and not isinstance(ical_date, datetime):
        return ical_date.isoformat()
    elif isinstance(ical_date, datetime):
        return ical_date.date().isoformat()
    else:
        raise ValueError(f"Invalid iCal date type: {type(ical_date)}")


def format_date_for_display(date_input: Union[str, datetime, date], format_type: str = 'short') -> str:
    """
    Format a date for display in UI templates.
    
    Args:
        date_input: Date string, datetime, or date object
        format_type: 'short' (Jan 15, 2024), 'long' (January 15, 2024), 'iso' (2024-01-15)
        
    Returns:
        Formatted date string
    """
    if date_input is None:
        return 'N/A'
    
    try:
        # Convert to date object
        if isinstance(date_input, str):
            if 'T' in date_input:
                # Parse datetime string and extract date
                dt = datetime.fromisoformat(date_input.replace('Z', '+00:00'))
                date_obj = dt.date()
            else:
                # Parse date string
                date_obj = datetime.strptime(date_input, '%Y-%m-%d').date()
        elif isinstance(date_input, datetime):
            date_obj = date_input.date()
        elif isinstance(date_input, date):
            date_obj = date_input
        else:
            return 'Invalid Date'
        
        # Format based on type
        if format_type == 'short':
            return date_obj.strftime('%b %d, %Y')  # Jan 15, 2024
        elif format_type == 'long':
            return date_obj.strftime('%B %d, %Y')  # January 15, 2024
        elif format_type == 'iso':
            return date_obj.isoformat()  # 2024-01-15
        else:
            return date_obj.strftime('%b %d, %Y')  # Default to short
            
    except (ValueError, TypeError):
        return 'Invalid Date'


def normalize_reservation_dates(reservation_data: dict) -> dict:
    """
    Normalize reservation date fields to consistent format.
    
    This function handles the various field names used across different
    parts of the application and ensures consistent date-only format.
    
    Args:
        reservation_data: Dictionary containing reservation data
        
    Returns:
        Dictionary with normalized date fields
    """
    normalized = reservation_data.copy()
    
    # Map of possible field names to standard names
    date_field_mappings = {
        'startDate': ['startDate', 'StartDate', 'checkInDate', 'CheckInDate', 'start'],
        'endDate': ['endDate', 'EndDate', 'checkOutDate', 'CheckOutDate', 'end']
    }
    
    for standard_field, possible_fields in date_field_mappings.items():
        date_value = None
        
        # Find the first available field
        for field in possible_fields:
            if field in reservation_data and reservation_data[field]:
                date_value = reservation_data[field]
                break
        
        if date_value:
            try:
                normalized[standard_field] = to_date_only(date_value)
            except ValueError as e:
                print(f"Warning: Could not normalize {standard_field}: {e}")
                normalized[standard_field] = None
        else:
            normalized[standard_field] = None
    
    return normalized


def is_date_in_range(check_date: Union[str, datetime, date], 
                    start_date: Union[str, datetime, date], 
                    end_date: Union[str, datetime, date]) -> bool:
    """
    Check if a date falls within a date range (inclusive).
    
    Args:
        check_date: Date to check
        start_date: Start of range
        end_date: End of range
        
    Returns:
        True if check_date is within the range
    """
    try:
        check_date_str = to_date_only(check_date)
        start_date_str = to_date_only(start_date)
        end_date_str = to_date_only(end_date)
        
        return start_date_str <= check_date_str <= end_date_str
    except (ValueError, TypeError):
        return False


def get_current_date_string() -> str:
    """
    Get current date as a date-only string.
    
    Returns:
        Current date in YYYY-MM-DD format
    """
    return datetime.now(timezone.utc).date().isoformat()


def ensure_date_only_format(date_input: Union[str, datetime, date, None]) -> Optional[str]:
    """
    Ensure date is in date-only format (YYYY-MM-DD) regardless of input format.
    
    This function is specifically designed to handle reservation dates
    and prevent timezone issues by always returning date-only strings.
    
    Args:
        date_input: Any date input (datetime, date, string, or None)
        
    Returns:
        Date string in YYYY-MM-DD format or None if input is invalid/None
    """
    if date_input is None:
        return None
    
    try:
        return to_date_only(date_input)
    except (ValueError, TypeError):
        return None


def format_date_for_ui(date_input: Union[str, datetime, date, None]) -> str:
    """
    Format date specifically for UI display with consistent formatting.
    
    This ensures that dates display the same way regardless of user timezone
    by treating all reservation dates as date-only values.
    
    Args:
        date_input: Date to format
        
    Returns:
        Formatted date string for display (e.g., "Jan 15, 2024") or "Invalid Date"
    """
    if date_input is None:
        return "Invalid Date"
    
    try:
        date_only_str = ensure_date_only_format(date_input)
        if not date_only_str:
            return "Invalid Date"
        
        # Parse the date-only string and format for display
        date_obj = datetime.strptime(date_only_str, '%Y-%m-%d').date()
        return date_obj.strftime('%b %d, %Y')  # Jan 15, 2024
    except (ValueError, TypeError):
        return "Invalid Date"


def is_reservation_active(start_date: Union[str, datetime, date], 
                         end_date: Union[str, datetime, date]) -> bool:
    """
    Check if a reservation is currently active (guest is staying).
    
    Args:
        start_date: Reservation start date
        end_date: Reservation end date
        
    Returns:
        True if reservation is active today
    """
    today = get_current_date_string()
    return is_date_in_range(today, start_date, end_date)


def is_reservation_upcoming(start_date: Union[str, datetime, date]) -> bool:
    """
    Check if a reservation is upcoming (starts in the future).
    
    Args:
        start_date: Reservation start date
        
    Returns:
        True if reservation starts in the future
    """
    try:
        today = get_current_date_string()
        start_date_str = to_date_only(start_date)
        return start_date_str > today
    except (ValueError, TypeError):
        return False

File: concierge/static/js/date_utils.js
/*
 * Date utilities for consistent date handling across the frontend
 * 
 * This module provides functions to handle reservation dates consistently
 * and avoid timezone-related display issues between hosts and guests.
 */

/**
 * Parse a date string safely, treating it as a date-only value
 * 
 * @param {string|null|undefined} dateStr - Date string in YYYY-MM-DD format or ISO datetime
 * @returns {Date|null} - Date object or null if invalid
 */
function parseDateSafely(dateStr) {
    if (!dateStr || typeof dateStr !== 'string') {
        return null;
    }

    try {
        // If the date string contains time info, extract just the date part
        let dateOnlyStr = dateStr;
        if (dateStr.includes('T')) {
            dateOnlyStr = dateStr.split('T')[0];
        }

        // Validate format (YYYY-MM-DD)
        if (!/^\d{4}-\d{2}-\d{2}$/.test(dateOnlyStr)) {
            return null;
        }

        // Create date object treating it as local date (avoid timezone conversion)
        const [year, month, day] = dateOnlyStr.split('-').map(Number);
        return new Date(year, month - 1, day); // month is 0-indexed
    } catch (e) {
        console.warn('Error parsing date:', dateStr, e);
        return null;
    }
}

/**
 * Format a date for display, consistently showing the same date regardless of user timezone
 * 
 * @param {string|Date|null|undefined} dateInput - Date to format
 * @param {string} format - Format type: 'short' (Jan 15, 2024), 'long' (January 15, 2024), 'iso' (2024-01-15)
 * @returns {string} - Formatted date string
 */
function formatDateForDisplay(dateInput, format = 'short') {
    if (!dateInput) {
        return 'Invalid Date';
    }

    let dateObj;
    if (typeof dateInput === 'string') {
        dateObj = parseDateSafely(dateInput);
    } else if (dateInput instanceof Date) {
        dateObj = dateInput;
    } else {
        return 'Invalid Date';
    }

    if (!dateObj) {
        return 'Invalid Date';
    }

    try {
        const options = {
            timeZone: 'UTC' // This ensures consistent display regardless of user timezone
        };

        switch (format) {
            case 'short':
                return dateObj.toLocaleDateString('en-US', {
                    ...options,
                    month: 'short',
                    day: 'numeric',
                    year: 'numeric'
                }); // Jan 15, 2024
            case 'long':
                return dateObj.toLocaleDateString('en-US', {
                    ...options,
                    month: 'long',
                    day: 'numeric',
                    year: 'numeric'
                }); // January 15, 2024
            case 'iso':
                return dateObj.toISOString().split('T')[0]; // 2024-01-15
            default:
                return dateObj.toLocaleDateString('en-US', {
                    ...options,
                    month: 'short',
                    day: 'numeric',
                    year: 'numeric'
                }); // Default to short
        }
    } catch (e) {
        console.warn('Error formatting date:', dateInput, e);
        return 'Invalid Date';
    }
}

/**
 * Check if a reservation is currently active (guest is staying)
 * 
 * @param {string} startDate - Start date in YYYY-MM-DD format
 * @param {string} endDate - End date in YYYY-MM-DD format
 * @returns {boolean} - True if reservation is active today
 */
function isReservationActive(startDate, endDate) {
    if (!startDate || !endDate) {
        return false;
    }

    try {
        const start = parseDateSafely(startDate);
        const end = parseDateSafely(endDate);
        const today = new Date();
        
        // Reset time to compare dates only
        today.setHours(0, 0, 0, 0);
        
        if (!start || !end) {
            return false;
        }

        return start <= today && today <= end;
    } catch (e) {
        console.warn('Error checking reservation status:', startDate, endDate, e);
        return false;
    }
}

/**
 * Check if a reservation is upcoming (starts in the future)
 * 
 * @param {string} startDate - Start date in YYYY-MM-DD format
 * @returns {boolean} - True if reservation starts in the future
 */
function isReservationUpcoming(startDate) {
    if (!startDate) {
        return false;
    }

    try {
        const start = parseDateSafely(startDate);
        const today = new Date();
        
        // Reset time to compare dates only
        today.setHours(0, 0, 0, 0);
        
        if (!start) {
            return false;
        }

        return start > today;
    } catch (e) {
        console.warn('Error checking upcoming reservation:', startDate, e);
        return false;
    }
}

/**
 * Get reservation status badge class and text
 * 
 * @param {string} startDate - Start date in YYYY-MM-DD format
 * @param {string} endDate - End date in YYYY-MM-DD format
 * @returns {Object} - Object with badge class and text
 */
function getReservationStatus(startDate, endDate) {
    if (isReservationActive(startDate, endDate)) {
        return {
            class: 'bg-success',
            text: 'Active'
        };
    } else if (isReservationUpcoming(startDate)) {
        return {
            class: 'bg-info',
            text: 'Upcoming'
        };
    } else {
        return {
            class: 'bg-secondary',
            text: 'Past'
        };
    }
}

/**
 * Sort reservations by date priority (active first, then upcoming, then past)
 * 
 * @param {Array} reservations - Array of reservation objects
 * @returns {Array} - Sorted array of reservations
 */
function sortReservationsByDate(reservations) {
    return reservations.sort((a, b) => {
        const aStart = parseDateSafely(a.startDate);
        const aEnd = parseDateSafely(a.endDate);
        const bStart = parseDateSafely(b.startDate);
        const bEnd = parseDateSafely(b.endDate);

        // Handle invalid dates
        if (!aStart || !aEnd) return 1;
        if (!bStart || !bEnd) return -1;

        const today = new Date();
        today.setHours(0, 0, 0, 0);

        // Determine status for each reservation
        const aActive = aStart <= today && today <= aEnd;
        const bActive = bStart <= today && today <= bEnd;
        const aUpcoming = aStart > today;
        const bUpcoming = bStart > today;

        // Active reservations first
        if (aActive && !bActive) return -1;
        if (!aActive && bActive) return 1;

        // Then upcoming reservations
        if (aUpcoming && !bUpcoming) return -1;
        if (!aUpcoming && bUpcoming) return 1;

        // Within same category, sort by date
        if (aUpcoming && bUpcoming) {
            // For upcoming, sooner comes first
            return aStart - bStart;
        } else {
            // For past, more recent comes first
            return bEnd - aEnd;
        }
    });
}

// Export functions for use in other modules
window.DateUtils = {
    parseDateSafely,
    formatDateForDisplay,
    isReservationActive,
    isReservationUpcoming,
    getReservationStatus,
    sortReservationsByDate
}; 